File: numass-main/src/main/java/inr/numass/models/sterile/NumassResolution.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package inr.numass.models.sterile;

import hep.dataforge.context.Context;
import hep.dataforge.maths.MathPlugin;
import hep.dataforge.meta.Meta;
import hep.dataforge.stat.parametric.AbstractParametricBiFunction;
import hep.dataforge.values.NamedValueSet;
import inr.numass.models.ResolutionFunction;
import inr.numass.utils.ExpressionUtils;
import org.apache.commons.math3.analysis.BivariateFunction;

import java.util.HashMap;
import java.util.Map;

import static java.lang.Math.sqrt;

/**
 * @author <a href="mailto:altavir@gmail.com">Alexander Nozik</a>
 */
public class NumassResolution extends AbstractParametricBiFunction {

    private static final String[] list = {}; //leaving 

    private final double resA;
    private double resB = 0;
    private BivariateFunction tailFunction = ResolutionFunction.getConstantTail();

    public NumassResolution(Context context, Meta meta) {
        super(list);
        this.resA = meta.getDouble("A", 8.3e-5);
        this.resB = meta.getDouble("B", 0);
        if (meta.hasValue("tail")) {
            String tailFunctionStr = meta.getString("tail");
            if (tailFunctionStr.startsWith("function::")) {
                tailFunction = MathPlugin.buildFrom(context).buildBivariateFunction(tailFunctionStr.substring(10));
            } else {
                tailFunction = (E, U) -> {
                    Map<String, Object> binding = new HashMap<>();
                    binding.put("E", E);
                    binding.put("U", U);
                    binding.put("D", E - U);
                    return ExpressionUtils.function(tailFunctionStr, binding);
                };
            }
        } else if (meta.hasValue("tailAlpha")) {
            //add polynomial function here
            double alpha = meta.getDouble("tailAlpha");
            double beta = meta.getDouble("tailBeta", 0);
            tailFunction = (double E, double U) -> 1 - (E - U) * (alpha + E / 1000d * beta) / 1000d;

        } else {
            tailFunction = ResolutionFunction.getConstantTail();
        }
    }

    @Override
    public double derivValue(String parName, double x, double y, NamedValueSet set) {
        return 0;
    }

    private double getValueFast(double E, double U) {
        double delta = resA * E;
        if (E - U < 0) {
            return 0;
        } else if (E - U > delta) {
            return tailFunction.value(E, U);
        } else {
            return (E - U) / delta;
        }
    }

    @Override
    public boolean providesDeriv(String name) {
        return true;
    }

    @Override
    public double value(double E, double U, NamedValueSet set) {
        assert resA > 0;
        if (resB <= 0) {
            return this.getValueFast(E, U);
        }
        assert resB > 0;
        double delta = resA * E;
        if (E - U < 0) {
            return 0;
        } else if (E - U > delta) {
            return tailFunction.value(E, U);
        } else {
            return (1 - sqrt(1 - (E - U) / E * resB)) / (1 - sqrt(1 - resA * resB));
        }
    }

}

File: numass-main/src/main/kotlin/inr/numass/models/sterile/NumassResolution.kt
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package inr.numass.models.sterile

import hep.dataforge.context.Context
import hep.dataforge.maths.functions.FunctionLibrary
import hep.dataforge.meta.Meta
import hep.dataforge.stat.parametric.AbstractParametricBiFunction
import hep.dataforge.values.Values
import inr.numass.models.ResolutionFunction
import inr.numass.utils.ExpressionUtils
import org.apache.commons.math3.analysis.BivariateFunction
import java.lang.Math.sqrt
import java.util.*

/**
 * @author [Alexander Nozik](mailto:altavir@gmail.com)
 */
class NumassResolution(context: Context, meta: Meta) : AbstractParametricBiFunction(list) {

    private val resA: Double = meta.getDouble("A", 8.3e-5)
    private val resB = meta.getDouble("B", 0.0)
    private val tailFunction: BivariateFunction = when {
        meta.hasValue("tail") -> {
            val tailFunctionStr = meta.getString("tail")
            if (tailFunctionStr.startsWith("function::")) {
                FunctionLibrary.buildFrom(context).buildBivariateFunction(tailFunctionStr.substring(10))
            } else {
                BivariateFunction { E, U ->
                    val binding = HashMap<String, Any>()
                    binding["E"] = E
                    binding["U"] = U
                    binding["D"] = E - U
                    ExpressionUtils.function(tailFunctionStr, binding)
                }
            }
        }
        meta.hasValue("tailAlpha") -> {
            //add polynomial function here
            val alpha = meta.getDouble("tailAlpha")
            val beta = meta.getDouble("tailBeta", 0.0)
            BivariateFunction { E: Double, U: Double -> 1 - (E - U) * (alpha + E / 1000.0 * beta) / 1000.0 }

        }
        else -> ResolutionFunction.getConstantTail()
    }

    override fun derivValue(parName: String, x: Double, y: Double, set: Values): Double {
        return 0.0
    }

    private fun getValueFast(E: Double, U: Double): Double {
        val delta = resA * E
        return when {
            E - U < 0 -> 0.0
            E - U > delta -> tailFunction.value(E, U)
            else -> (E - U) / delta
        }
    }

    override fun providesDeriv(name: String): Boolean {
        return true
    }

    override fun value(E: Double, U: Double, set: Values): Double {
        assert(resA > 0)
        if (resB <= 0) {
            return this.getValueFast(E, U)
        }
        assert(resB > 0)
        val delta = resA * E
        return when {
            E - U < 0 -> 0.0
            E - U > delta -> tailFunction.value(E, U)
            else -> (1 - sqrt(1 - (E - U) / E * resB)) / (1 - sqrt(1 - resA * resB))
        }
    }

    companion object {

        private val list = arrayOf<String>() //leaving
    }

}
